src/coreclr/jit/llvm.cpp

```diff
@@ -892,100 +894,44 @@ Value* Llvm::consumeValue(GenTree* node, Type* targetLlvmType)
 
-    // TODO-LLVM: delete this when PUT_ARG inserted for the return address
-    thisArgNum          = 0;
-    if (needsReturnStackSlot(sigInfo.retType, sigInfo.retTypeClass))
-    {
-        // return slot before this
-        thisArgNum++;
-        returnAddressArgNum = 0;
-    }
+    std::vector<llvm::Value*> argVec = std::vector<llvm::Value*>();
 
     for (unsigned i = 0; i < argCount; i++)
     {
         fgArgTabEntry* curArgTabEntry = argTable[i];
-        unsigned int   argNum         = curArgTabEntry->argNum;
-        OperandArgNum  opAndArg       = {argNum, curArgTabEntry->GetNode()};
-        sortedData[argNum]            = opAndArg;
-    }
-
-    for (OperandArgNum opAndArg : sortedArgs)
-    {
-        if (opAndArg.operand->IsArgPlaceHolderNode())
+        if (curArgTabEntry->GetNode()->IsArgPlaceHolderNode())
         {
-            // TODO-LLVM: delete place holder args when rewriting the signature
-            thisArgNum++;
-            returnAddressArgNum++;
             continue;
         }
 
-        // "this" not in sigInfo arg list, dont increment argIx.  TODO-LLVM look to remove when PUTARG is inserted
-        if (opAndArg.argNum == thisArgNum && sigInfo.hasThis())
-        {
-            storeOnShadowStack(opAndArg.operand, shadowStackForCallee, shadowStackUseOffest);
-            shadowStackUseOffest += TARGET_POINTER_SIZE;
-            continue;
-        }
+        unsigned int argNum = curArgTabEntry->argNum;
 
-        // Return address arg not in sigInfo, but needs to be added to LLVM args. TODO-LLVM delete when PUT_ARG is inserted
-        if (returnAddressArgNum == opAndArg.argNum)
-        {
-            argVec.push_back(consumeValue(opAndArg.operand, Type::getInt8PtrTy(_llvmContext)));
-            continue;
-        }
+        assert(curArgTabEntry->GetNode()->OperIs(GT_PUTARG_TYPE));
+        GenTreePutArgType* putArg = curArgTabEntry->GetNode()->AsPutArgType();
+        argVec.push_back(consumeValue(putArg->gtGetOp1(), getLlvmTypeForCorInfoType(putArg->gtCorInfoType, putArg->gtClsHnd)));
+    }
 
-        LlvmArgInfo llvmArgInfo = getLlvmArgInfoForArgIx(sigInfo, argIx);
-        if (llvmArgInfo.m_argIx >= 0)
-        {
-            // pass the parameter on the LLVM stack
-            argVec.push_back(consumeValue(opAndArg.operand, llvmFunc->getArg(llvmArgInfo.m_argIx)->getType()));
-        }
-        else
-        {
-            // pass on shadow stack
-            storeOnShadowStack(opAndArg.operand, shadowStackForCallee, shadowStackUseOffest);
-            shadowStackUseOffest += TARGET_POINTER_SIZE;
-        }
-        argIx++;
+    if (llvmFunc == nullptr)
+    {
+        CORINFO_ARG_LIST_HANDLE sigArgs = sigInfo.args;
+
+        // assume ExternalLinkage, if the function is defined in the clrjit module, then it is replaced and an extern
+        // added to the Ilc module
+        llvmFunc = Function::Create(getFunctionTypeForSigInfo(call->gtCallIrReturnType, sigInfo),
```

Comment:
Hmm, the idea was that we'd determine the `FunctionType` from the arguments in IR and only them.

Reply:
Thanks for bringing back on track!   Yes, this was fairly straightforward.  I do still need something on the `GenTreeCall` so hold the true return type, but with that this becomes `getFunctionTypeForCall(call)` and 
```
FunctionType* Llvm::getFunctionTypeForCall(GenTreeCall* call)
{
    llvm::Type* retLlvmType = getLlvmTypeForCorInfoType(call->gtCallIrReturnType, call->gtRetClsHnd);

    std::vector<llvm::Type*> argVec = std::vector<llvm::Type*>();

    for (GenTreeCall::Use& use : call->Args())
    {
        GenTreePutArgType* putArg = use.GetNode()->AsPutArgType();
        argVec.push_back(getLlvmTypeForCorInfoType(putArg->gtCorInfoType, putArg->gtClsHnd));
    }

    return FunctionType::get(retLlvmType, ArrayRef<Type*>(argVec), false);
}
```
I'll rename `gtCallIrReturnType`